### DISCUSSION TITLE
How can I specify the default format for time outputs?

### DISCUSSION THREAD

**Discussion by Willard21** [josdejong/mathjs#3050] [Q&A]
If I evaluate an expression like `"17 seconds * 20000"`, then print it with `toString()`, ideally it would return `"3 days 22 hours 26 minutes 40 seconds"` or something to that effect. Instead, it returns `"340 kiloseconds"`, which isn't the most useful time format.

If an output unit is specified then obviously it shouldn't print the time string. The exact format could optionally be configurable, but anything human-readable would be fine.

**Comment by josdejong** [comment]
Thanks for your suggestion Will, makes sense to see if we can output time in better human readable format 😄 

Ideally, the output would be also valid input to pass to the expression parser, this is so far the case with all formatted outputs which is very powerful I think. Your example `3 days 22 hours 26 minutes 40 seconds` would be difficult in that regard: it will be interpreted as implicit multiplications. I'm not sure if there is a way around that.

A simple solution would be to automatically select a better prefix, so the output would be `3.9351851851852 days`. 

Or maybe we could introduce a special `formatTime` function which could get all freedom and could actually output `3 days 22 hours 26 minutes 40 seconds`. If we go that way, it may be easier/better to leave this up to a library which is specialized in formatting time instead of reinventing the wheel.

**Comment by Willard21** [comment]
Perhaps printing it as `"(3 days + 22 hours + 26 minutes + 40 seconds)"` then? It's perfectly readable, and can still be correctly parsed by the expression parser. Personally I use this library to parse arbitrary input from users, and I love how it just works out-of-the box without me needing to know anything about the type of data being parsed.

3.9351851851852 days is decent too, and is honestly what I expected to see when I first used it. The full breakdown of times would be better for my use case (since I'm lazy), but I wouldn't want it to break anything else in the process.

Perhaps you could add a new unit called "duration" that's represented as seconds internally, but expands into the human readable format when toString() is called on it. I dunno if that would work, but it's an idea.

**Comment by josdejong** [comment]
> Perhaps printing it as "(3 days + 22 hours + 26 minutes + 40 seconds)" then

wow that's so simple and obvious, good idea 👍 .

It may be best to create a separate function like `formatTime` to output `(3 days + 22 hours + 26 minutes + 40 seconds)`, and improve the default behavior of `evaluate('17 seconds * 20000').toString()` so that it outputs `3.9351851851852 days` instead of `340 kiloseconds`, so this is consistent with all other units.

> Perhaps you could add a new unit called "duration" that's represented as seconds internally, but expands into the human readable format when toString() is called on it. I dunno if that would work, but it's an idea.

Hm, interesting idea. I'm not sure whether it would work out nicely, maybe it will only get more confusing. I have the feeling this is only about the _formatting_ at the end of the calculations, and not at calculation level itself, so I would say we should find a solution at formatting level.

**Comment by gwhitney** [comment]
What the original question asks for can (at least at this point) be easily accomplished by using `Unit.splitUnit()`, which takes an array of units and chops a Unit value into pieces accordingly, returning a corresponding area of the parts as Unit values. For example,
```
const u = math.multiply(math.unit('17 seconds'), 20000)
console.log('A long time', u.toString())
console.log('Still long', u.splitUnit(['days', 'hours', 'minutes', 'seconds']).map(p => p.toString()))
```
produces on the console
```
A long time 340 kiloseconds
Still long [ '3 days', '22 hours', '26 minutes', '40 seconds' ]
```

Given this, I will rephrase this as a question and move it to Discussions Q&A.
